# Configuring the Pipeline for a Complete Run

This document explains how to correctly configure the pipeline for a complete run. For a minimal example which documents basic configuration along with pipeline execution, refer to [Quick Start: A Minimal Example](./minimal_example.md).

- [Configuring the Pipeline for a Complete Run](#configuring-the-pipeline-for-a-complete-run)
  - [Complete Example Configuration File](#complete-example-configuration-file)
  - [NCBI FCS Adaptor](#ncbi-fcs-adaptor)
  - [NCBI FCS GX](#ncbi-fcs-gx)
  - [BUSCO](#busco)
  - [TIDK](#tidk)
  - [LAI](#lai)
  - [KRAKEN2](#kraken2)
  - [HIC](#hic)

## Complete Example Configuration File

This document explains the pipeline configuration using an example configuration file packaged with the pipeline. Refer to 'conf/test_full.config'. This configuration is an expansion of the 'conf/test_minimal.config' covered in [Quick Start: A Minimal Example](./minimal_example.md).

## NCBI FCS Adaptor

This module has only one parameter: `empire`. The permissible values are: `euk` for Eukaryotes and `prok` for Prokaryotes.

> From conf/test_full.config

```groovy
ncbi_fcs_adaptor {
    empire = 'euk'
}
```

## NCBI FCS GX

Following parameters must be configured:

- `tax_id`: The taxonomy ID for all the target assemblies listed in the `target_assemblies` parameter. A taxonomy ID can be obtained by searching a *Genus species* at <https://www.ncbi.nlm.nih.gov/taxonomy>. A single ID for all assemblies implies that the pipeline is designed to be used for QC'ing one or assemblies of the same *species* in one run.
- `db_manifest_url`: This URL specifies the database version used by the pipeline.
- `db_path`: This is the path to the database files stored on a directory accessible to the pipeline. The data placed inside this directory should match with the `db_manifest_url`. Otherwise, the pipeline with an error. Before running the pipeline, the user must make sure that the data is correctly downloaded and placed in a directory accessible to the pipeline. Setup instructions are available at <https://github.com/ncbi/fcs/wiki/FCS-GX>. The database directory should contain following files:

```bash
all.assemblies.tsv
all.blast_div.tsv.gz
all.gxi
all.gxs
all.manifest
all.meta.jsonl
all.README.txt
all.seq_info.tsv.gz
all.taxa.tsv
```

> From conf/test_full.config

```groovy
ncbi_fcs_gx {
    tax_id          = "35717"
    db_manifest_url = "https://ftp.ncbi.nlm.nih.gov/genomes/TOOLS/FCS/database/r2023-01-24/all.manifest"
    db_path         = "/workspace/ComparativeDataSources/NCBI/FCS/GX/r2023-01-24"
}
```

## BUSCO

Following parameters must be configured:

- `mode`: geno or genome, for genome assemblies (DNA), tran or transcriptome, for transcriptome assemblies (DNA); and prot or proteins, for annotated gene sets (protein).
- `lineage_datasets`: A list of BUSCO lineages. Any number of lineages can be specified. Each target assembly is assessed against each of the listed lineage. To select a lineage, refer to <https://busco.ezlab.org/list_of_lineages.html>
- `download_path`: A directory where the BUSCO can download and cache its databases. BUSCO manages download and validation of the databases itself, there, fore, the user does not have to manually setup these databases.

> From conf/test_full.config

```groovy
busco {
    mode              = "geno"
    lineage_datasets  = ["fungi_odb10", "hypocreales_odb10"]
    download_path     = "/workspace/ComparativeDataSources/BUSCO/assembly_qc"
}
```

## TIDK

Following parameters must be configured:

- repeat_seq: The telomere search sequence. To select an appropriate sequence, see <http://telomerase.asu.edu/sequences_telomere.html>. Commonly used sequences are: TTTAGGG (Plant), TTAGGG (Fungus, Vertebrates), TTAGG (Insect).
- filter_by_size: Set this flag to 1 to filter out assembly sequences smaller than the size specified by the next parameter (default: 0).
- filter_size_bp: Minimum size of the assembly sequence processed by TIDK (default: 1000000 (1Mbp)).

> From conf/test_full.config

```groovy
tidk {
    repeat_seq = "TTAGGG"
}
```

In the example configuration above, the `filter_by_size` and `filter_size_bp` are not set. The pipeline will pick up their default values from 'nextflow.config' file.

## LAI

Following parameters must be configured:

- mode: "" for Standard, "-q" for Quick and "-qq" for Very Quick. The Very Quick mode only produces Raw LAI which is only valid for intra-species comparisons.
- pass_list: A list of lists which specifies the `*pass.list` file for each assembly. This file is generated by EDTA or LTR_retriever software. If this parameter is `null`, the pipeline first runs EDTA and itself generates the `*.pass.list` file for each assembly. If `*.pass.list` file for each assembly is available, specify it in the `pass_list` parameter along with the tag for the corresponding target assembly. Default value for this parameter is `null`. Here is an example:

```groovy
target_assemblies   = [
    ["hap1", "/workspace/assembly_qc/test_data/default/test_data1.fasta.gz"],
    ["hap2", "/workspace/assembly_qc/test_data/default/test_data2.fasta"]
]

lai {
    pass_list       = [
        ["hap1", "/workspace/assembly_qc/test_data/default/test_data1.pass.list"],
        ["hap2", "/workspace/assembly_qc/test_data/default/test_data2.pass.list"]
    ]
}
```

Notice that the tags (hap1, hap2) in target_assemblies and pass_list are matched.

- out_file: Similar to `*.pass.list`, an `*.out` file is also required for LAI calculation. The specification rules for `out_file` are exactly same as those for `pass_list`. Default value for this parameter is `null`. Here is an example:

```groovy
target_assemblies   = [
    ["hap1", "/workspace/assembly_qc/test_data/default/test_data1.fasta.gz"],
    ["hap2", "/workspace/assembly_qc/test_data/default/test_data2.fasta"]
]

lai {
    pass_list       = [
        ["hap1", "/workspace/assembly_qc/test_data/default/test_data1.pass.list"],
        ["hap2", "/workspace/assembly_qc/test_data/default/test_data2.pass.list"]
    ]

    out_file        = [
        ["hap1", "/workspace/assembly_qc/test_data/default/test_data1.out"],
        ["hap2", "/workspace/assembly_qc/test_data/default/test_data2.out"]
    ]
}
```

> WARNING: EDTA modifies the sequence names in a fasta file if they are not short (<=13 characters) and simple (i.e, letters, numbers, and underscore; no comments allowed in the ID filed). When specifying the `*pass.list` and `*.out` files, make sure that these files have the same sequence IDs as those in the fasta files specified by the `target_assemblies` parameter.

- edta::is_sensitive: "--sensitive" parameter for the EDTA software. Set to 1 to turn on Sensitive (very slow) and 0 to turn off Sensitive. Default is 0. For Tair10, the LAI score with Sensitive turned off is 17.90 and with Sensitive turned on is 17.87.

> From conf/test_full.config

```groovy
lai {
    mode              = "" // Standard
    pass_list         = null
    out_file          = null

    edta {
        is_sensitive  = 0
    }
}
```

Notice that the default values are used for all the LAI parameters. There is no need to re-define these parameters if the default values are used. It is done here for the sake of clarity.

## KRAKEN2

Following parameters must be configured:

- db_url: The URL for the KRAKEN2 database. To select a DB, see <https://benlangmead.github.io/aws-indexes/k2>. By default, the pipeline uses PlusPFP database.
- download_path: The directory where the pipeline can download and cache the data. The pipeline can download the database itself. However, care must be taken when switching databases. If the `db_url` is changed, a new `download_path` should be provided, otherwise, the pipeline will fail with error saying that the downloaded database is corrupted.

> From conf/test_full.config

```groovy
kraken2 {
    db_url        = "https://genome-idx.s3.amazonaws.com/kraken/k2_pluspfp_20230314.tar.gz"
    download_path = "/workspace/ComparativeDataSources/kraken2db/k2_pluspfp_20230314"
}
```

## HIC

Following parameters must be configured:

- paired_reads: A relative or absolute path to paired reads in fastq.gz format, or a SRA ID. The format for file path is `*R{1,2}*.(fasta|fq).gz`. An example is '/input/genomic/fungal/Neonectria/Genome/20190506_CAGRF19591_CGYCF_HiC/PG_PETUNIA_HiC_CGYCF_CACTCA_L001_R{1,2}.fastq.gz'.

> From conf/test_full.config

```groovy
hic {
    paired_reads = "SRR8238190"
}
```

To be continued...
